main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+# the size is unordered, can be a list or also a set
+
+print(A.shape)
+# prints { width[:64], height[:64], channel[:3] }
+\end{python}
+
+We can also convert a standard  tensor `T` to a named one.
+ 
+\begin{python}
+A = namedtensor.tensor(T,[width, height]) 
+# first axis becomes width and second height.
+# lengths are inherited from T and so do not need to be specified.
+\end{python}
+
+
+If we define new functions for named tensors, we can use something like
+
+\begin{python}
+  @signature(X=[width,height,channel],W=[width,height,channel], output=[width,height])
+  def CONV(X,W):
+      Y = namedtensor.tensor({(x,y) : sum((a,b) in zip(width(W),height(W)) X[x+a,y+b]*W[a,b]} for (x,y) in zip(width(X),height(X))))  
+\end{python}
+We can use assertions to require some conditions on the lengths of the different axes.
+
+
+
+We will also have a way in Python to extend existing functions and modules that were written for unnamed tensors.
+Maybe something like
+
+\begin{python}
+g = named(X=[width,height], output = [layer] ,f)
+\end{python}
+when \pyth{f} is such a function that takes \pyth{X} as a tensor input. In this case width will be mapped to the first axis and height to the second one. 
+
+Should also have such way to automatically transform pytorch modules of the form `nn.Module` to ones that use named tensors as their weights, inputs, and outputs.
+
+As usual, the signature of an operator does not specify all the information about the relation of its input and output relation. For example, we can have the "halving" operator $HALVE$ that drops the bottom half of an image.
+
+$HALVE:\mathbb{R}^{\name{width},\name{height}} \rightarrow \mathbb{R}^{\name{width},\name{height}}$
+
+The abstract types of the input and output tensors are the same but we can add the constraint that $|width(HALVE(X))|=\lfloor |width(X)/2| \rfloor$ separately.
+
+
+
+\subsection{Example}
+
+
+If $\name{foo}, \name{bar}$ are axes (supported over the natural numbers) then the following is a tensor of shape $\{ \name{foo}[..2], \name{bar}[..3] \}$ and type $\{ \name{foo} , \name{bar} \}$: 
+
+\begin{equation*}
+A = \nmatrix{foo}{bar}{
+  3 & 1 & 4 \\
+  1 & 5 & 9
+}.
+\end{equation*}
+
+
+$A$ is a map from $Ind \{ \name{foo}[..2] , \name{bar}[..3] \}$ to $\mathbb{R}$. Since order does not matter, we can also think of it as a map from $\name{foo}[..2] \times \name{bar}[..3]$ to $\mathbb{R}$.
+So, given $i \in name{foo}[..2]$ and $j \in \name{bar}[..3]$, we can write $A(i,j)$ or $A_{i,j}$ for the corresponding element of $A$.
+In particular $A_{\name{foo}[1],\name{bar}[3]}=4$. We can also write this as $A_{\name{foo}=1,\name{bar}=4}$. In Python-speak, while this is not how we implement tennsors, we could think of $A$ as a function that takes two parameters, named \pyth{foo} and \pyth{bar} and outputs a real number. The type of the parameter \pyth{foo} is \pyth{foo} and the type of the parameter \pyth{bar} is \pyth{bar}.
```

Comment:
This analogy is not quite right; it's more like Python keyword arguments, because both have the requirement that names are unique.

Reply:
Yes I meant keyword arguments - will fix this